crates/sui/src/client_commands.rs

```diff
@@ -1021,6 +1026,103 @@ impl SuiClientCommands {
+
+    let read_api = client.read_api();
+
+    let sui_system_state = read_api.get_sui_system_state().await.unwrap();
+    let mut max = 0;
+    let validators = &sui_system_state.validators.active_validators;
+    // We don't have pending_validators now
+    for val in validators {
+        let name: &str = std::str::from_utf8(&val.metadata.name).unwrap();
+        let new_total_stake = val.metadata.next_epoch_stake + val.metadata.next_epoch_delegation;
+        if max < new_total_stake {
+            max = new_total_stake;
+        }
+        println!(
+            "{:<20}: next epoch stake {:.3} SUI",
+            name,
+            new_total_stake / 1_000_000_000
+        );
+    }
+    println!("max: {max}");
+
+    let coins = context
+        .gas_objects(signer)
+        .await
+        .unwrap()
+        .iter()
+        .map(|q| GasCoin::try_from(&q.1).unwrap())
+        .collect::<Vec<GasCoin>>();
+
+    if coins.len() < (validators.len() - 1) * 2 {
+        panic!("Not enough coins!");
+    }
+    let mut coins = coins.iter();
+
+    let mut txes = Vec::with_capacity(validators.len());
+    for val in &sui_system_state.validators.active_validators {
+        let name: &str = std::str::from_utf8(&val.metadata.name).unwrap();
+        let address = val.metadata.sui_address;
+        let new_total_stake = val.metadata.next_epoch_stake + val.metadata.next_epoch_delegation;
+        let diff = max - new_total_stake;
+        if diff > 0 {
+            println!(
+                "{:<20}: Planning to stake {:.3} SUI",
+                name,
+                diff / 1_000_000_000
+            );
+            let delegate_coin = coins.find(|c| c.value() >= diff).unwrap();
+            let gas_coin = coins.find(|c| c.value() >= GAS_BUDGET).unwrap();
+            let tx = client
+                .transaction_builder()
+                .request_add_delegation(
+                    signer,
+                    vec![*delegate_coin.id()],
+                    Some(diff),
+                    // Some(1),
+                    address,
+                    Some(*gas_coin.id()),
+                    GAS_BUDGET,
+                )
+                .await
+                .unwrap();
+            let signature = context
+                .config
+                .keystore
+                .sign_secure(&signer, &tx, Intent::default())
+                .unwrap();
+            txes.push((
+                name,
+                Transaction::from_data(tx, Intent::default(), signature)
+                    .verify()
+                    .unwrap(),
+            ));
+        }
+    }
+
+    for (name, tx) in txes {
+        let mut cnt = 1;
+        let tx_digst = *tx.digest();
+        loop {
+            if let Err(err) = context.execute_transaction(tx.clone()).await {
+                println!(
+                    "{:<20}: {:?} failed for {} time(s): {:?}",
+                    name, tx_digst, cnt, err
+                );
```

Comment:
Potential infinite loop if the txn fails to execute due to something deterministic like move aborts?

Reply:
But other than that the logic seems fine